File: typescript/packages/extensions/src/sign-in-with-x/sign.ts
/**
 * Message signing for SIWX extension
 *
 * Client-side helpers for signing SIWX messages.
 * Supports both EVM (viem) and Solana wallet adapters.
 */

import { encodeBase58 } from "./solana";

/**
 * Signer interface for EVM SIWX message signing.
 * Compatible with viem WalletClient and PrivateKeyAccount.
 */
export interface EVMSigner {
  /** Sign a message and return hex-encoded signature */
  signMessage: (args: { message: string; account?: unknown }) => Promise<string>;
  /** Account object (for WalletClient) */
  account?: { address: string };
  /** Direct address (for PrivateKeyAccount) */
  address?: string;
}

/**
 * Wallet adapter style Solana signer.
 * Compatible with @solana/wallet-adapter, Phantom/Solflare wallet APIs.
 */
export interface WalletAdapterSigner {
  /** Sign a message and return raw signature bytes */
  signMessage: (message: Uint8Array) => Promise<Uint8Array>;
  /** Solana public key (Base58 encoded string or PublicKey-like object) */
  publicKey: string | { toBase58: () => string };
}

/**
 * Solana Kit KeyPairSigner style.
 * Compatible with createKeyPairSignerFromBytes and generateKeyPairSigner from @solana/kit.
 */
export type SolanaKitSigner = {
  /** Solana address (Base58 encoded string) */
  address: string;
  /** Sign messages - accepts messages with content and signatures */
  signMessages: (
    messages: Array<{ content: Uint8Array; signatures: Record<string, unknown> }>,
  ) => Promise<Array<Record<string, Uint8Array>>>;
};

/**
 * Union type for Solana signers - supports both wallet adapter and @solana/kit.
 */
export type SolanaSigner = WalletAdapterSigner | SolanaKitSigner;

/**
 * Union type for SIWX signers - supports both EVM and Solana wallets.
 */
export type SIWxSigner = EVMSigner | SolanaSigner;

/**
 * Get address from an EVM signer.
 *
 * @param signer - EVM wallet signer instance
 * @returns The wallet address as a hex string
 */
export function getEVMAddress(signer: EVMSigner): string {
  if (signer.account?.address) {
    return signer.account.address;
  }
  if (signer.address) {
    return signer.address;
  }
  throw new Error("EVM signer missing address");
}

/**
 * Get address from a Solana signer.
 * Supports both wallet adapter (publicKey) and @solana/kit (address) interfaces.
 *
 * @param signer - Solana wallet signer instance
 * @returns The wallet address as a Base58 string
 */
export function getSolanaAddress(signer: SolanaSigner): string {
  // Check for @solana/kit KeyPairSigner interface (address property)
  if ("address" in signer && signer.address) {
    return signer.address;
  }
  // Fall back to wallet adapter interface (publicKey property)
  if ("publicKey" in signer) {
    const pk = signer.publicKey;
    return typeof pk === "string" ? pk : pk.toBase58();
  }
  throw new Error("Solana signer missing address or publicKey");
}

/**
 * Sign a message with an EVM wallet.
 * Returns hex-encoded signature.
 *
 * @param message - The message to sign
 * @param signer - EVM wallet signer instance
 * @returns Hex-encoded signature
 */
export async function signEVMMessage(message: string, signer: EVMSigner): Promise<string> {
  if (signer.account) {
    return signer.signMessage({ message, account: signer.account });
  }
  return signer.signMessage({ message });
}

/**
 * Sign a message with a Solana wallet.
 * Returns Base58-encoded signature.
 * Supports both wallet adapter (signMessage) and @solana/kit (signMessages) interfaces.
 *
 * @param message - The message to sign
 * @param signer - Solana wallet signer instance
 * @returns Base58-encoded signature
 */
export async function signSolanaMessage(message: string, signer: SolanaSigner): Promise<string> {
  const messageBytes = new TextEncoder().encode(message);

  // Check for @solana/kit signMessages interface
  if ("signMessages" in signer) {
    const results = await signer.signMessages([{ content: messageBytes, signatures: {} }]);
    // signMessages returns an array of signature dictionaries
    // The signature is keyed by the signer's address
    const sigDict = results[0] as { [key: string]: Uint8Array };
    // Get the first (and only) signature value from the dictionary
    const signatureBytes = Object.values(sigDict)[0];
    return encodeBase58(signatureBytes);
  }

  // Fall back to wallet adapter signMessage interface
  if ("signMessage" in signer) {
    const signatureBytes = await signer.signMessage(messageBytes);
    return encodeBase58(signatureBytes);
  }

  throw new Error("Solana signer missing signMessage or signMessages method");
}

File: python/x402/extensions/sign_in_with_x/sign.py
"""Message signing for SIWX extension."""

from __future__ import annotations

from typing import Any, Protocol

from .solana import encode_base58


class EVMSigner(Protocol):
    """Signer interface for EVM SIWX message signing."""

    async def sign_message(self, *, message: str, account: Any = None) -> str: ...


class SolanaSigner(Protocol):
    """Solana wallet signer (adapter or solders Keypair wrapper)."""

    async def sign_message(self, message: bytes) -> bytes: ...


SIWxSigner = Any


def get_evm_address(signer: Any) -> str:
    """Get address from an EVM signer."""
    account = getattr(signer, "account", None)
    if account is not None and getattr(account, "address", None):
        return account.address
    address = getattr(signer, "address", None)
    if isinstance(address, str):
        return address
    raise ValueError("EVM signer missing address")


def get_solana_address(signer: Any) -> str:
    """Get address from a Solana signer."""
    address = getattr(signer, "address", None)
    if isinstance(address, str) and address:
        return address
    public_key = getattr(signer, "public_key", None) or getattr(signer, "publicKey", None)
    if public_key is None:
        raise ValueError("Solana signer missing address or publicKey")
    if isinstance(public_key, str):
        return public_key
    if hasattr(public_key, "to_base58"):
        return public_key.to_base58()
    if hasattr(public_key, "toBase58"):
        return public_key.toBase58()
    raise ValueError("Solana signer missing address or publicKey")


async def sign_evm_message(message: str, signer: Any) -> str:
    """Sign a message with an EVM wallet and return hex-encoded signature."""
    try:
        from eth_account.messages import encode_defunct
    except ImportError as e:
        raise ImportError(
            "SIWX EVM signing requires eth-account. Install with: pip install x402[evm]"
        ) from e

    account = getattr(signer, "_account", None) or getattr(signer, "account", None)
    if account is None and hasattr(signer, "sign_message") and hasattr(signer, "address"):
        account = signer

    if account is not None and hasattr(account, "sign_message"):
        signed = account.sign_message(encode_defunct(text=message))
        raw = signed.signature
        return raw if isinstance(raw, str) and raw.startswith("0x") else "0x" + raw.hex()

    if hasattr(signer, "sign_message") and not hasattr(signer, "signMessage"):
        sig = signer.sign_message(message=message, account=getattr(signer, "account", None))
        if hasattr(sig, "__await__"):
            sig = await sig
        return _normalize_evm_signature(sig)

    raise ValueError("EVM signer missing sign_message support")


def _normalize_evm_signature(sig: Any) -> str:
    if hasattr(sig, "signature"):
        raw = sig.signature
        if isinstance(raw, bytes):
            return "0x" + raw.hex()
        return str(raw)
    if isinstance(sig, bytes):
        return "0x" + sig.hex()
    return str(sig)


async def sign_solana_message(message: str, signer: Any) -> str:
    """Sign a message with a Solana wallet and return Base58-encoded signature."""
    message_bytes = message.encode("utf-8")

    if hasattr(signer, "sign_messages") and callable(signer.sign_messages):
        results = signer.sign_messages([{"content": message_bytes, "signatures": {}}])
        if hasattr(results, "__await__"):
            results = await results
        sig_dict = results[0]
        signature_bytes = next(iter(sig_dict.values()))
        return encode_base58(signature_bytes)

    if hasattr(signer, "sign_message"):
        sig = signer.sign_message(message_bytes)
        if hasattr(sig, "__await__"):
            sig = await sig
        return encode_base58(bytes(sig))

    try:
        from solders.keypair import Keypair
    except ImportError as e:
        raise ImportError(
            "SIWX Solana signing requires solders. Install with: pip install x402[svm]"
        ) from e

    keypair = getattr(signer, "keypair", None)
    if isinstance(keypair, Keypair):
        return encode_base58(bytes(keypair.sign_message(message_bytes)))

    if isinstance(signer, Keypair):
        return encode_base58(bytes(signer.sign_message(message_bytes)))

    raise ValueError("Solana signer missing sign_message or sign_messages method")
